apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py

```diff
@@ -102,70 +165,60 @@ def prepare(element_html, data):
     pl.check_answers_names(data, answer_name)
 
     check_indentation = pl.get_boolean_attrib(element, "indentation", INDENTION_DEFAULT)
-    grading_method = pl.get_string_attrib(
-        element, "grading-method", GRADING_METHOD_DEFAULT
+    grading_method = pl.get_enum_attrib(
+        element, "grading-method", GradingMethodType, GRADING_METHOD_DEFAULT
+    )
+    feedback_type = pl.get_enum_attrib(
+        element, "feedback", FeedbackType, FEEDBACK_DEFAULT
     )
-    feedback_type = pl.get_string_attrib(element, "feedback", FEEDBACK_DEFAULT)
 
-    if grading_method in ["dag", "ranking"]:
-        partial_credit_type = pl.get_string_attrib(element, "partial-credit", "lcs")
-        if partial_credit_type not in ["none", "lcs"]:
-            raise Exception(
-                'partial credit type "'
-                + partial_credit_type
-                + '" is not available with the "'
-                + grading_method
-                + '" grading-method.'
-            )
-    elif pl.get_string_attrib(element, "partial-credit", None) is not None:
+    if (
+        grading_method is not GradingMethodType.DAG
+        and grading_method is not GradingMethodType.RANKING
+        and pl.get_enum_attrib(element, "partial-credit", PartialCreditType) is not None
```

Comment:
@eliotwrobson I edited the `get_enum_attrib` implementation because this call was causing a crash with `"Attribute "partial-credit" missing and no default is available'`

`get_enum_attrib` should allow there to be no default provided, and to return `None` in that case, because that's how `get_string_attrib` et al. behave.

Reply:
Oh wait, I don't think they allow no default, but they do allow `None` to be passed as default.  Anyway, I had to change things to get it working. If I did something wrong let me know.

Reply:
This was actually a very deliberate choice when I first implemented this function. If none is now allowed to be a default value, then that needs to be expressed by changing the type annotation, which will trigger errors unless fixed. The new annotation has to be more complex (requires an overload). But I was actually skeptical whether it was ever necessary to use `None` as a return value, or whether that instead encouraged inconsistent programming habits.

In this case, you set the default value to `None`, but only use it to check whether the default is returned (meaning the attribute was not set). But instead, it's more expressive just to check whether element has partial-credit set as an attribute at all (i.e. with `pl.has_attrib(element, "partial-credit")`). So you can replace the `is None` check here with that has attribute check, and the `get_enum_attrib` function should be left unchanged. The ideal thing would actually be to see if it were possible to disallow returning a sentinel for all of these getter functions, as it would mean not having to use the overloaded type annotations.

Reply:
You make a compelling argument for leaving it, and you are right that using `has_attrib` makes more sense here either way. I'll go ahead and do that here for now.

I still think it is less-than-ideal that now the `pl.get_*_attrib` functions have inconsistent behavior, and we should talk about how to resolve this in the longer term.